File: examples/javascript/closures.js
// makeCounter returns a function that increments a private counter on each call.
const makeCounter = function (start) {
  let n = start;
  return function () {
    n += 1;
    return n;
  };
};

// identity returns its single argument unchanged (block-bodied arrow).
const identity = (x) => {
  return x;
};

// pair builds a two-element array (expression-bodied arrow).
const pair = (a, b) => [a, b];

module.exports = { makeCounter, identity, pair };

File: examples/golang/closures.go
package examples

// MakeAdder returns a closure that adds x to its argument.
func MakeAdder(x int) func(int) int {
	return func(y int) int {
		return x + y
	}
}

// ApplyTwice applies f to v two times using a nested helper closure.
func ApplyTwice(v int, f func(int) int) int {
	step := func(n int) int {
		return f(n)
	}
	return step(step(v))
}
